packages/gatsby/src/schema/node-model.js

```diff
@@ -963,6 +963,9 @@ const addRootNodeToInlineObject = (
   const isPlainObject = _.isPlainObject(data)
 
   if (isPlainObject || _.isArray(data)) {
+    if (isNode && data.internal.dontTrackInlineObjects) {
+      return
+    }
```

Comment:
this should probably be better in https://github.com/gatsbyjs/gatsby/blob/66ed27ab37baa175ce993eb8dbb7f48ee2ba1d99/packages/gatsby/src/schema/node-model.js#L500-L511 as that's entry point to run internal `addRootNodeToInlineObject` function - we don't need to do those checks on each depth level and `node` argument there is ~guaranteed to be node anyway so it should be able to skip `isNode`-like check there